rand_distr/src/weighted_tree.rs

```diff
@@ -0,0 +1,364 @@
+            *weight > W::ZERO
+        } else {
+            false
+        }
+    }
+
+    /// Gets the weight at an index.
+    pub fn get(&self, index: usize) -> W
+    where
+        W: for<'a> SubAssign<&'a W>,
+    {
+        let left_index = 2 * index + 1;
+        let right_index = 2 * index + 2;
+        let mut w = self.subtotals[index].clone();
+        w -= &self.subtotal(left_index);
+        w -= &self.subtotal(right_index);
+        w
+    }
+
+    /// Removes the last weight and returns it, or [`None`] if it is empty.
+    pub fn pop(&mut self) -> Option<W>
+    where
+        W: for<'a> SubAssign<&'a W>,
+    {
+        self.subtotals.pop().map(|weight| {
+            let mut index = self.len();
+            while index != 0 {
+                index = (index - 1) / 2;
+                self.subtotals[index] -= &weight;
+            }
+            weight
+        })
+    }
+
+    /// Appends a new weight at the end.
+    pub fn push(&mut self, weight: W) -> Result<(), WeightedError> {
+        if weight < W::ZERO {
+            return Err(WeightedError::InvalidWeight);
+        }
+        if let Some(total) = self.subtotals.first() {
+            let mut total = total.clone();
+            if total.checked_add_assign(&weight).is_err() {
+                return Err(WeightedError::Overflow);
+            }
+        }
+        let mut index = self.len();
+        self.subtotals.push(weight.clone());
+        while index != 0 {
+            index = (index - 1) / 2;
+            self.subtotals[index].checked_add_assign(&weight).unwrap();
+        }
+        Ok(())
+    }
+
+    /// Updates the weight at an index.
+    pub fn update(&mut self, mut index: usize, weight: W) -> Result<(), WeightedError>
+    where
+        W: for<'a> SubAssign<&'a W>,
+    {
+        if weight < W::ZERO {
+            return Err(WeightedError::InvalidWeight);
+        }
+        let mut difference = weight;
+        difference -= &self.get(index);
+        if difference == W::ZERO {
+            return Ok(());
+        }
+        if let Some(total) = self.subtotals.first() {
+            let mut total = total.clone();
+            if total.checked_add_assign(&difference).is_err() {
+                return Err(WeightedError::Overflow);
+            }
+        }
+        self.subtotals[index]
+            .checked_add_assign(&difference)
+            .unwrap();
+        while index != 0 {
+            index = (index - 1) / 2;
+            self.subtotals[index]
+                .checked_add_assign(&difference)
+                .unwrap();
+        }
+        Ok(())
+    }
+
+    fn subtotal(&self, index: usize) -> W {
+        if index < self.subtotals.len() {
+            self.subtotals[index].clone()
+        } else {
+            W::ZERO
+        }
+    }
+}
+
+impl<W: Clone + PartialEq + PartialOrd + SampleUniform + SubAssign<W> + Weight>
+    Distribution<Result<usize, WeightedError>> for WeightedTreeIndex<W>
```

Comment:
We don't usually implement `Distribution<Result<..>>`. Do we *need* to?

I think I would prefer to panic on error, but guarantee no panic if `self.is_valid()` (`self.can_sample()`).

@vks?